src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -90,72 +93,93 @@ internal Type GetTypeFromCoreAssembly(CoreTypeId typeId)
             return null;
         }
 
-        internal void AppendMetadata(MetadataBuilder metadata)
+        internal void AppendMetadata()
         {
             // Add module metadata
-            metadata.AddModule(
+            _metadataBuilder.AddModule(
                 generation: 0,
-                moduleName: metadata.GetOrAddString(_name),
-                mvid: metadata.GetOrAddGuid(Guid.NewGuid()),
+                moduleName: _metadataBuilder.GetOrAddString(_name),
+                mvid: _metadataBuilder.GetOrAddGuid(Guid.NewGuid()),
                 encId: default,
                 encBaseId: default);
 
             // Create type definition for the special <Module> type that holds global functions
-            metadata.AddTypeDefinition(
+            _metadataBuilder.AddTypeDefinition(
                 attributes: default,
                 @namespace: default,
-                name: metadata.GetOrAddString("<Module>"),
+                name: _metadataBuilder.GetOrAddString("<Module>"),
                 baseType: default,
                 fieldList: MetadataTokens.FieldDefinitionHandle(1),
                 methodList: MetadataTokens.MethodDefinitionHandle(1)); ;
 
             // Add each type definition to metadata table.
-            foreach (TypeBuilderImpl typeBuilder in _typeDefStore)
+            foreach ((TypeBuilderImpl typeBuilder, TypeDefinitionHandle typeHandle) in _typeDefinitions)
```

Comment:
`_typeDefinitions` cannot be dictionary. Enumeration of dictionary does not guarantee order. It has to be List, so that the order is guaranteed.

Reply:
Good point, thanks!